CONTRIBUTING.md

```diff
@@ -189,6 +190,81 @@ library maintainers on the exclusion from this rule. In such cases there is a
 requirement to test unsafe code with sanitizers including Miri.
 
 
+### Policy
+
+We have various `rust-bitcoin` specific coding styles and conventions that are
+grouped here loosely under the term 'policy'. These are things we try to adhere
+to but that you should not need to worry too much about if you are a new
+contributor. Think of this as a place to collect group knowledge that exists in
+the various PRs over the last few years.
+
+
+#### Errors
+
+Return as much context as possible with errors e.g., if an error was encountered parsing a string
+include the string in the returned error type. If a function consumes costly-to-compute input
+(allocations are also considered costly) it should return the input back in the error type.
+
+More specifically errors should
+
+- be `non_exhaustive`
+- have private fields unless you're very confident they won't change.
+- only derive  `Debug`
+- implement Display using `write_err!()` macro if a variant contains an inner error source.
+- have `Error` suffix
+- implement `std::error::Error` if they are public (feature gated on "std").
+- commit to Send and Sync use `pub_error_type!` macro
+
+```rust
+/// Documentation for the `Error` type.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Documentation for variant A.
+    A,
+    /// Documention for variant B.
+    B,
+}
+pub_errro_type!(Error)
+```
+
+
+#### rustdocs
+
+Be liberal with references to BIPS or other documention, the aim is that devs can learn about
+Bitcoin by hacking on this codebase as opposed to having to learn about Bitcoin first and then start
+hacking on this codebase. Consider this format
+
+
+```rust
+pub struct BlockHeader {
+    /// Originally protocol version, but repurposed for soft-fork signaling.
+    ///
+    /// ### Relevant BIPs
+    ///
+    /// * [BIP 9 - Version bits with timeout and delay](https://github.com/bitcoin/bips/blob/master/bip-0009.mediawiki) (current usage)
+    /// * [BIP 34 - Block v2, Height in Coinbase](https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki)
+    pub version: i32,
+```
+
+Do use rustdoc subheadings. Do put an empty newline below each heading e.g.,
```

Comment:
In 8578505bf4c2f3f01116c3a97715e2bfd1f88c64:

Maybe we should also say that you are required to add a `Panics` section if the function panics under any circumstances (including array indexing on input). For `Errors` it's probably ok to let this be a judgement call since some functions have "obvious" error returns.

Reply:
Oh and we could mention `track_caller`  (your mention of array index out of bound triggered this thought).